docs: end to end testing writing tests. (#14671)
gatsbyjs/gatsby: docs/docs/end-to-end-testing.md

---

> title: End-to-end testing

Reviewer: ```suggestion
title: End-to-End Testing
```

While we're updating this, let's make it adhere to the Style Guide.

Author: Sure. Do you have more suggestions or should I commit this?

Reviewer: I do, but I had to jump into a meeting. Stay tuned

---
title: End-to-end testing
---

> [Cypress](https://www.cypress.io/) is one of the options when it comes to end-to-end (E2E) testing. Cypress is an all-in-one testing framework focused on E2E testing meaning that you don't have to install 10 different things to get your test suite set up. You can write your first passing test in minutes without any configuration with the help of Cypress' API which is easy to read and understand. It runs tests as fast as your browser can render content which also makes test-driven development possible. You'll also profit from the time travel feature or the extensive debugging capabilities with Chrome DevTools. Of course you can also use it with Gatsby and this guide will explain how.

Reviewer 2: ```suggestion
[Cypress](https://www.cypress.io/) is one of the options when it comes to end-to-end (E2E) testing. Cypress is an all-in-one testing framework focused on E2E testing, meaning that you don't have to install 10 different things to get your test suite set up. You can write your first passing test in minutes without any configuration with the help of Cypress' API, which is easy to read and understand. It runs tests as fast as your browser can render content, which also makes test-driven development possible. You'll also profit from the time travel feature or the extensive debugging capabilities with Chrome DevTools. Of course you can also use it with Gatsby, and this guide will explain how.
```

A few commas would improve the readability of this paragraph.

npm install --save-dev cypress start-server-and-test
```

> We also want the URLs used by `cy.visit()` or `cy.request()` to be prefixed hence you have to create the file `cypress.json` at the root of your project with the following content:

Reviewer 2: ```suggestion
We also want the URLs used by `cy.visit()` or `cy.request()` to be prefixed, hence you have to create the file `cypress.json` at the root of your project with the following content:
```

  "scripts": {
    "develop": "gatsby develop",
    "cy:open": "cypress open",
    "cy:run": "cypress run", // highlight-line
    "test:e2e": "start-server-and-test develop http://localhost:8000 cy:open"
    "test:e2e:ci": "start-server-and-test develop http://localhost:8000 cy:run" // highlight-line
  }
}
```

Please read the [Cypress' official documentation](https://docs.cypress.io/guides/guides/continuous-integration.html) on CI if you want to know how to setup Travis or GitLab with Cypress.

## Writing tests

> The first test is an **accessibility** test with [cypress-axe](https://github.com/avanslaars/cypress-axe).

Reviewer: ```suggestion
A good use case for writing automated end-to-end tests is asserting **accessibility** with [cypress-axe](https://github.com/avanslaars/cypress-axe), a Cypress plugin that incorporates the [axe](https://deque.com/axe) accessibility testing API. While some [manual testing](https://www.smashingmagazine.com/2018/09/importance-manual-accessibility-testing/) is still required to ensure good web accessibility, automation can ease the burden on human testers.
```
I think this needs to be introduced a bit more generally. It also provides an opportunity to discuss the both the benefits and limits of accessibility automation.

  "scripts": {
    "develop": "gatsby develop",
    "cy:open": "cypress open",
    "cy:run": "cypress run", // highlight-line
    "test:e2e": "start-server-and-test develop http://localhost:8000 cy:open"
    "test:e2e:ci": "start-server-and-test develop http://localhost:8000 cy:run" // highlight-line
  }
}
```

Please read the [Cypress' official documentation](https://docs.cypress.io/guides/guides/continuous-integration.html) on CI if you want to know how to setup Travis or GitLab with Cypress.

## Writing tests

> The first test is an **accessibility** test with [cypress-axe](https://github.com/avanslaars/cypress-axe).

Reviewer: I feel like I should know the answer to this, but what do you mean by "the first test"?

Author: I mean that is the first test someone will write by reading this page.

To use cypress-axe you have to install `cypress-axe` and [axe-core](https://github.com/dequelabs/axe-core). You'll also use some commands from [cypress-testing-library](https://testing-library.com/docs/cypress-testing-library/intro) to target elements easier:

```bash
npm install --save-dev cypress-axe axe-core @testing-library/cypress
```

Then you add the `cypress-axe` and `cypress-testing-library` commands in `cypress/support/commands.js`:

```js:title=cypress/support/commands.js
import "./commands"
//highlight-start
import "cypress-axe"
import "@testing-library/cypress/add-commands"
//highlight-end
```

Cypress right now will look for tests inside the `cypress/integration` folder. It makes sense to create a new folder `cypress/e2e` (end to end). You can use that folder for tests by adding the following in your `cypress.json`:

```json:title=cypress.json
{
  "baseUrl": "http://localhost:8000/"
  "integrationFolder": "cypress/e2e" // highlight-line
}
```

Create a new file inside `cypress/e2e` folder and name it `a11y.test.js`.

> You'll use the `beforeEach` hook to run some commands before each test. After cypress loads the homepage you'll use the `checkA11y` method to check for accessibility violations:

Reviewer: ```suggestion
You'll use the `beforeEach` hook to run some commands before each test. After cypress loads the homepage you'll use the `checkA11y` method from `cypress-axe` to check for accessibility violations:
```

```js:title=cypress/e2e/a11y.test.js
/// <reference types="Cypress" />

describe("Accessibility checks", () => {
  beforeEach(() => {
    cy.visit("/")
    cy.injectAxe()
    cy.wait(500)
  })
  it("Has no detectable a11y violations on load", () => {
    cy.checkA11y()
  })
})
```

You can run `test:e2e` to run the test. If you already have the development server open you can run `cy:open` instead.

> One thing to keep in mind is that you can't always see the exact error message from the sidebar (command log). For that, you have to open the developer console and find the message in the output. You can see how an accessibility error looks in the [cypress-axe GitHub page](https://github.com/avanslaars/cypress-axe#output).

Reviewer: ```suggestion
One thing to keep in mind is that you can't always see the exact error message from the sidebar (command log). For that, you have to open the browser developer console and find the message in the output. You can see how an accessibility error looks in the [cypress-axe GitHub page](https://github.com/avanslaars/cypress-axe#output).
```

You don't have to use the `checkA11y` method only on page load. For example, you can perform a click on a button and check again. This is especially useful if that button opens a modal or a mobile menu for example.

> The following test is for the [gatsby-default-starter](https://github.com/gatsbyjs/gatsby-starter-default). You visit the homepage and you search for the link that goes to page 2 with the `getByText` command. Then you perform a click and check for accessibility errors in the second page.

Reviewer: ```suggestion
The following tests are for the [gatsby-default-starter](https://github.com/gatsbyjs/gatsby-starter-default). The first test visits the homepage and runs `cypress-axe` on it. In the second test, Cypress searches for a link that goes to page 2 with the `getByText` command, performs a click event on the link, and checks for accessibility errors on the second page.
```
The user isn't doing the work, the automated test is...so this suggestion reflects that!

Author: > The user isn't doing the work, the automated test is.

Yes, I agree, thanks! 

My objection here is that there is only one test, the highlighted. I'm including the previous test just to show what changed in the test file.

    cy.checkA11y()
  })
  // highlight-start
  it("Navigates to page 2 and checks for accessibility violations", () => {
    cy.getByText(/go to page 2/i)
      .click()
      .checkA11y()
  })
  // highlight-end
})
```

You can also make [assertions](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Assertions) with the [should command](https://docs.cypress.io/api/commands/should.html#Syntax).

In `gatsby-default-starter` homepage you can write another test that focuses on the footer link. Then you make some assertions on the item that's currently focused:

```js:title=cypress/e2e/a11y.test.js
/// <reference types="Cypress" />

describe("Accessibility checks", () => {
  beforeEach(() => {
    cy.visit("/")
    cy.injectAxe()
    cy.wait(500)
  })
  it("Has no detectable a11y violations on load", () => {
    cy.checkA11y()
  })
  it("Navigates to page 2 and checks for accessibility violations", () => {
    cy.getByText(/go to page 2/i)
      .click()
      .checkA11y()
  })
  // highlight-start
  it("Checks if footer link is focusable and has the correct attributes", () => {
    cy.getAllByText("Gatsby")
      .focus()

>       .checkA11y()

Reviewer: Cypress-axe doesn't do anything for to verify focusability or color contrast of focus styles, so running cypress-axe on it here probably isn't useful. You could potentially scope `cypress-axe` to just the footer, though, by using axe.run parameters: https://github.com/avanslaars/cypress-axe#parameters-on-cychecka11y-axerun

Author: Maybe just remove the `checkA11y` then. I didn't find something that we could use by changing the `axe.run` parameters. Only the scope you mention (to make the test run faster?).

    cy.focused()
      .should("have.text", "Gatsby")
      .should("have.attr", "href", "https://www.gatsbyjs.org")

>       .should("have.css", "outline-width", "1px")

Author: This line fails when I run it with `cypress run` (it pass with `cypress open`). It says the `outline-width` is `5px` for some reason.

I will change it to `.should("not.have.css", "outline-width", "0px")` to ensure that at least there's an outline.
